File: 2-linear-optimization/matlab/ex_assignment.m
% ex_assignment.m
% example of an assignment network problem where a company is planning to
% assign three people to three jobs
% [reference] Example 8.4 in LNO
% [course] Session 6 - Linear Optimization (3)
close all; clear; clc

% problem data
c = [11 5 2 15 12 8 3 1 10];
% c = [11 5 2 20 12 8 3 1 10]; % alternative c vector
A = []; % not present
b = []; % not present
Aeq = [1 1 1 0 0 0 0 0 0;
    0 0 0 1 1 1 0 0 0;
    0 0 0 0 0 0 1 1 1;
    -1 0 0 -1 0 0 -1 0 0;
    0 -1  0 0 -1 0 0 -1 0;
    0 0 -1 0 0 -1 0 0 -1];
beq = [1 1 1 -1 -1 -1];
LB = zeros(length(c),1);
UB = ones(length(c),1);

% options
OPTIONS = optimoptions('linprog');
OPTIONS.Display = 'iter';

% solve using linprog
[X,FVAL,EXITFLAG,OUTPUT] = linprog(-c,A,b,Aeq,beq,LB,UB,OPTIONS)

%--------------------------------------------------------------------------
% visualize the solution
hf = figure; hf.Color = 'w'; hold on

% colors
niceblue = [77, 121, 167]/255;
nicered = [225, 86, 86]/255;
nicegreen = [109, 195, 80]/255;

% create the graph
s = [1 1 1 2 2 2 3 3 3];
t = [4 5 6 4 5 6 4 5 6];
weights = c;
names = {'Person 1','Person 2','Person 3',...
    'Accountant','Budget Director','Personnel Manager'};
G = digraph(s,t,weights,names);

% determine line widths
LWidths = 4*G.Edges.Weight/max(G.Edges.Weight);

% assign edge colors
EdgeColor = repmat([0 0 0],length(c),1);
EdgeColor(logical(X),:) = repmat(nicegreen,sum(X),1);

% create the figure
hp = plot(G,'LineWidth',LWidths,'Layout','layered',...
    'Direction','right','EdgeColor',EdgeColor);

% set node color
hp.NodeColor = nicered;
hp.NodeFontSize = 12;

% manually position nodes consistent with book picture
% XData = hp.XData;
% XData(1) = 18.709;
% XData(2) = 18.709;
% XData(3) = 18.709;
% XData(4) = 87.734;
% XData(5) = 87.734;
% XData(6) = 87.734;
% hp.XData = XData;
% 
% YData = hp.YData;
% YData(1) = 46.260;
% YData(2) = 25.103;
% YData(3) = 3.948;
% YData(4) = 46.260;
% YData(5) = 25.103;
% YData(6) = 3.948;
% hp.YData = YData;

axis off
axis equal

File: 2-linear-optimization/python/ex_assignment.py
import numpy as np 
from scipy.optimize import linprog
import networkx as nx
import matplotlib.pyplot as plt

# problem data 
c = np.array([11,5,2,15,12,8,3,1,10])

# equality constraints
Aeq = np.array([[1,1,1,0,0,0,0,0,0],
    [0,0,0,1,1,1,0,0,0],
    [0,0,0,0,0,0,1,1,1],
    [-1,0,0,-1,0,0,-1,0,0],
    [0,-1,0,0,-1,0,0,-1,0],
    [0,0,-1,0,0,-1,0,0,-1]])
beq = np.array([1,1,1,-1,-1,-1])

# bounds
LB = np.zeros((len(c),))
UB = np.ones((len(c),))

# combine bounds and reshape
bounds = np.vstack([LB,UB]); bounds = bounds.T

# specify method 
method = 'highs'

# linprog options
options = {'disp':True}

# solve the optimization problem
res = linprog(-c,A_eq = Aeq,b_eq = beq,bounds = bounds,method = method, options = options)

# create directed graph
G = nx.DiGraph()
G.add_node('Person 1')
G.add_node('Person 2')
G.add_node('Person 3')

G.add_node('Personal Manager')
G.add_node('Budget Director')
G.add_node('Accountant')

node_list = ['Person 1','Person 2','Person 3']

edge_list = [('Person 1','Personal Manager'), ('Person 1','Budget Director'), ('Person 1','Accountant'),
                  ('Person 2','Personal Manager'), ('Person 2','Budget Director'), ('Person 2','Accountant'),
                  ('Person 3','Personal Manager'),('Person 3','Budget Director'),('Person 3','Accountant')]

G.add_edges_from(edge_list)

# create a circular layout
pos = nx.circular_layout(G)
edge_labels = dict(zip(edge_list,list(c)))

# plot the graph
nx.draw(G,pos = pos,with_labels = True)

nx.draw_networkx_edges(
    G,
    pos,
    edgelist = [('Person 3','Personal Manager'),('Person 2','Budget Director'),('Person 1','Accountant')],
    width = 3,
    alpha = 0.5,
    edge_color = 'tab:red')

nx.draw_networkx_edge_labels(
    G,
    pos,
    edge_labels = edge_labels
    )

plt.show()
